packages/compose/src/hooks/use-focus-outside/index.js

```diff
@@ -0,0 +1,191 @@
+ *
+ * @return {FocusOutsideReturnValue} An object containing event handlers. Bind the event handlers
+ *                                   to a wrapping element element to capture when focus moves
+ *                                   outside that element.
+ */
+export default function useFocusOutside( onFocusOutside ) {
+	const currentOnFocusOutside = useRef( onFocusOutside );
+	useEffect( () => {
+		currentOnFocusOutside.current = onFocusOutside;
+	}, [ onFocusOutside ] );
+
+	const preventBlurCheck = useRef( false );
+
+	/**
+	 * @type {import('react').MutableRefObject<number | undefined>}
+	 */
+	const blurCheckTimeoutId = useRef();
+
+	/**
+	 * Cancel a blur check timeout.
+	 */
+	const cancelBlurCheck = useCallback( () => {
+		clearTimeout( blurCheckTimeoutId.current );
+	}, [] );
+
+	// Cancel blur checks on unmount.
+	useEffect( () => {
+		return () => cancelBlurCheck();
+	}, [] );
+
+	// Cancel a blur check if the callback or ref is no longer provided.
+	useEffect( () => {
+		if ( ! onFocusOutside ) {
+			cancelBlurCheck();
+		}
+	}, [ onFocusOutside, cancelBlurCheck ] );
+
+	/**
+	 * Handles a mousedown or mouseup event to respectively assign and
+	 * unassign a flag for preventing blur check on button elements. Some
+	 * browsers, namely Firefox and Safari, do not emit a focus event on
+	 * button elements when clicked, while others do. The logic here
+	 * intends to normalize this as treating click on buttons as focus.
+	 *
+	 * @see https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button#Clicking_and_focus
+	 *
+	 * @param {SyntheticEvent} event Event for mousedown or mouseup.
+	 */
+	const normalizeButtonFocus = useCallback( ( event ) => {
+		const { type, target } = event;
+		const isInteractionEnd = includes( [ 'mouseup', 'touchend' ], type );
+
+		if ( isInteractionEnd ) {
+			preventBlurCheck.current = false;
+		} else if ( isFocusNormalizedButton( target ) ) {
+			preventBlurCheck.current = true;
+		}
+	}, [] );
+
+	/**
+	 * A callback triggered when a blur event occurs on the element the handler
+	 * is bound to.
+	 *
+	 * Calls the `onFocusOutside` callback in an immediate timeout if focus has
+	 * move outside the bound element and is still within the document.
+	 *
+	 * @param {SyntheticEvent} event Blur event.
+	 */
+	const queueBlurCheck = useCallback( ( event ) => {
+		// React does not allow using an event reference asynchronously
+		// due to recycling behavior, except when explicitly persisted.
+		event.persist();
+
+		// Skip blur check if clicking button. See `normalizeButtonFocus`.
+		if ( preventBlurCheck.current ) {
+			return;
+		}
+
+		blurCheckTimeoutId.current = setTimeout( () => {
+			// If document is not focused then focus should remain
+			// inside the wrapped component and therefore we cancel
+			// this blur event thereby leaving focus in place.
+			// https://developer.mozilla.org/en-US/docs/Web/API/Document/hasFocus.
+			if ( ! document.hasFocus() ) {
+				event.preventDefault();
```

Comment:
I don't think you can prevent default an event after a timeout?

Reply:
Good point! I did some digging and it looks like Andrew made a similar comment on the PR that introduced this, focus events aren't cancelable so `preventDefault` would have no effect - https://github.com/WordPress/gutenberg/pull/17051#discussion_r392391759

I'll remove this line in a separate PR.